beluga/include/beluga/sensor/data/cache_friendly_grid_storage_4.hpp

```diff
@@ -0,0 +1,164 @@
+
+  [[nodiscard]] static constexpr bool is_power_of_four(std::size_t n) { return is_power_of_base(n, 4); }
+
+  [[nodiscard]] static constexpr bool is_a_square_power_of_two(std::size_t n) {
+    // if it's a power of four, it's also a perfect square
+    return is_power_of_four(n);
+  }
+
+  [[nodiscard]] static constexpr std::size_t calculate_tile_area(std::size_t elements_per_line) {
+    if (is_a_square_power_of_two(elements_per_line)) {
+      // the elements in a line can be mapped to a square tile (4, 16, 64,
+      // etc) use a single line for each square tile
+      return elements_per_line;
+    }
+    // the elements in a line can't be mapped to a square tile (2, 8, 32, etc), so
+    // we use two consecutive lines for each tile
+    return elements_per_line * 2;
+  }
+
+  [[nodiscard]] static constexpr std::size_t calculate_tile_side(std::size_t elements_per_line) {
+    auto tile_area = calculate_tile_area(elements_per_line);
+    // we calculate the side length by calculating the square root base four,
+    // and elevating two to that power
+    std::size_t tile_side = 1;
+    while (tile_area > 1) {
+      tile_area /= 4;
+      tile_side *= 2;
+    }
+    return tile_side;
+  }
+
+  [[nodiscard]] static constexpr std::size_t ceil_div(std::size_t numerator, std::size_t denominator) {
+    return ((numerator + denominator - 1) / denominator);
+  }
+
+  [[nodiscard]] std::size_t mapping_function(std::size_t x, std::size_t y) const {
+    const auto l2_tile_x = x / kL2TileSide;
+    const auto l2_tile_y = y / kL2TileSide;
+    auto l2_cell_x = x % kL2TileSide;
+    auto l2_cell_y = y % kL2TileSide;
+
+    if ((l2_tile_x + l2_tile_y) & 0x01) {
+      std::swap(l2_cell_x, l2_cell_y);
+    }
+
+    const auto l1_tile_x = l2_cell_x / kL1TileSide;
+    const auto l1_tile_y = l2_cell_y / kL1TileSide;
+    const auto l1_cell_x = l2_cell_x % kL1TileSide;
+    const auto l1_cell_y = l2_cell_y % kL1TileSide;
+
+    const auto l2_tile_offset = (l2_tile_y * kL2Tilecols_ + l2_tile_x) * kL2TileSize;
+    const auto l1_tile_offset = (l1_tile_y * kL1TileCols + l1_tile_x) * kL1TileSize;
+    const auto l1_cell_offset = l1_cell_y * kL1TileSide + l1_cell_x;
+    return l2_tile_offset + l1_tile_offset + l1_cell_offset;
+  }
+
+  [[nodiscard]] constexpr std::size_t map_coordinates_to_index(int x, int y) const {
+    return mapping_function(static_cast<std::size_t>(x), static_cast<std::size_t>(y));
+  }
+
+  static constexpr std::size_t kL1TileCols{2};
+  static constexpr std::size_t kL1TileRows{2};
+
+  static constexpr std::size_t kL2TileSize{kL1TileRows * kL1TileCols * calculate_tile_area(LineLenght / sizeof(T))};
+  static constexpr std::size_t kL2TileSide{kL1TileCols * calculate_tile_side(LineLenght / sizeof(T))};
+  static constexpr std::size_t kL1TileSize{calculate_tile_area(LineLenght / sizeof(T))};
+  static constexpr std::size_t kL1TileSide{calculate_tile_side(LineLenght / sizeof(T))};
```

Comment:
@glpuga nit: would this be equivalent to:
```suggestion
  static constexpr std::size_t kL1TileCols{2};
  static constexpr std::size_t kL1TileRows{2};
  static constexpr std::size_t kL1TileSize{calculate_tile_area(LineLenght / sizeof(T))};
  static constexpr std::size_t kL1TileSide{calculate_tile_side(LineLenght / sizeof(T))};
  
  static constexpr std::size_t kL2TileSize{kL1TileRows * kL1TileCols * kL1TileSize};
  static constexpr std::size_t kL2TileSide{kL1TileCols * kL1TileSide};
```
?

Reply:
Yes, that's ok.